### DISCUSSION TITLE
Counter App Multiple Functions code is not compiling for me

### DISCUSSION THREAD

**Discussion by markalanboyd** [ratatui-org/website#151] [Q&A] [answered]
Greetings - I am working my way through the book and I ran into an error with the code at the bottom of this page.

https://ratatui.rs/tutorial/counter-app/multiple-functions.html

I can open an issue and a PR for this, but I wanted to check first if it's just me. This is the error I'm getting:

```rust
error[E0107]: struct takes 0 generic arguments but 1 generic argument was supplied
   --> src/main.rs:12:31
    |
12  | pub type Frame<'a> = ratatui::Frame<'a, CrosstermBackend<std::io::Stderr>>;
    |                               ^^^^^     --------------------------------- help: remove this generic argument
    |                               |
    |                               expected 0 generic arguments
    |
note: struct defined here, with 0 generic parameters
   --> /Users/markboyd/.cargo/registry/src/index.crates.io-6f17d22bba15001f/ratatui-0.24.0/src/terminal.rs:557:12
    |
557 | pub struct Frame<'a> {
    |            ^^^^^

For more information about this error, try `rustc --explain E0107`.
error: could not compile `ratatui-counter-app` (bin "ratatui-counter-app") due to previous error
```

When I change this line:

```rust
pub type Frame<'a> = ratatui::Frame<'a, CrosstermBackend<std::io::Stderr>>;
```

to this:

```rust
pub type Frame<'a> = ratatui::Frame<'a>;
```

then it compiles and the app works as expected.

I'm a complete newbie to Rust, and some of this here is over my head, but if someone else can confirm this, could I make a PR for this issue?

If it is user error, I would love to know also.

As a sanity check, here's my `Cargo.toml`

```toml
[package]
name = "ratatui-counter-app"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
anyhow = "1.0.75"
crossterm = "0.27.0"
ratatui = "0.24.0"
```

Just in case this is relevant:

Mac Mini
Apple M2 Pro
Sonoma 14.0
rustc 1.73.0 (cc66ad468 2023-10-03)

PS: I just got to the Question section at the bottom of the page, and the number I put in `from_millis(num)` has no effect on the speed, which seems to be set by my Mac's key repeat rate. I can put in 1, 0, or 10_000 and it does nothing.

It even has a little pause before it repeats, just like it normally does in a text editor app. I figure this might just be a peculiarity of how it run on my machine and doesn't have anything to do with Ratatui, so I figured it was worth mentioning but not enough to start its own topic.

**Comment by kdheepak** [comment, marked as answer]
Code in the tutorial is using ratatui v0.23 but this breaking change happened in v0.24: https://ratatui.rs/highlights/v0.24.html#frame-no-more-generics-

A PR would be welcome!

**Reply by markalanboyd** [reply]
Ok, will do!

**Comment by kdheepak** [comment]
> PS: I just got to the Question section at the bottom of the page, and the number I put in from_millis(num) has no effect on the speed, which seems to be set by my Mac's key repeat rate. I can put in 1, 0, or 10_000 and it does nothing.

Does it not affect your CPU utilization? With a low number you should see a higher CPU utilization.

**Reply by markalanboyd** [reply]
> Does it not affect your CPU utilization? With a low number you should see a higher CPU utilization.

Is this the info that would show a difference? I took a sample of the terminal from Activity Monitor - I couldn't see an appreciable difference in the Activity Monitor itself but I figured maybe since I have an M2 it wasn't a hard enough process to tax it noticeably.

`from_millis(10000)`
```
Analysis of sampling zsh (pid 6249) every 1 millisecond
Process:         zsh [6249]
Path:            /bin/zsh
Load Address:    0x1040f8000
Identifier:      zsh
Version:         100
Code Type:       ARM64E
Platform:        macOS
Parent Process:  zsh (figterm) [6230]

Date/Time:       2023-11-10 08:26:54.752 +0100
Launch Time:     2023-11-09 18:08:50.839 +0100
OS Version:      macOS 14.0 (23A344)
Report Version:  7
Analysis Tool:   /usr/bin/sample

Physical footprint:         3025K
Physical footprint (peak):  3057K
```

`from_millis(1)`
```
Analysis of sampling zsh (pid 6249) every 1 millisecond
Process:         zsh [6249]
Path:            /bin/zsh
Load Address:    0x1040f8000
Identifier:      zsh
Version:         100
Code Type:       ARM64E
Platform:        macOS
Parent Process:  zsh (figterm) [6230]

Date/Time:       2023-11-10 08:28:42.624 +0100
Launch Time:     2023-11-09 18:08:50.839 +0100
OS Version:      macOS 14.0 (23A344)
Report Version:  7
Analysis Tool:   /usr/bin/sample

Physical footprint:         3025K
Physical footprint (peak):  3057K
Idle exit:                  untracked
```

**Reply by markalanboyd** [reply]
Nevermind - I just found the real process and it takes up 50% of my cpu lol. Odd though that it has no effect on key repeat rate.
